### ISSUE TITLE
`incorrectly-parenthesized-tuple-in-subscript` (`RUF031`) with `parenthesize-tuple-in-subscript = true` false positives on types outside of type annotations

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#13065]
the fix for #12758 only fixed the issue for type annotations, but the generic syntax for types can appear outside of type annotations, eg:

```py
foo: dict[str, int] = {"a": 1} # no error
foo = dict[str, int](a=1) # RUF031: Use parentheses for tuples in subscripts.
```
[playground](https://play.ruff.rs/0a3d1260-7a71-4c6f-867e-8d9c9c274b00)

i'm not sure how feasible it is to fix this without type checking, perhaps it could just assume any index access followed by a call is using a type

**Comment by dylwil3** [comment]
Neat! This is tricky. Ignoring subscripts followed by calls would lead to false negatives for folks using a dispatch pattern:

```python
dispatch = {(0,1): north, (1,0): east, (0,-1): south, (-1,0): west }
def walk(x,y,z):
    return dispatch[x,y](z)
```
Maybe there is some sneakier way of teasing out this generic type situation...

**Comment by dhruvmanila** [comment]
Yeah, I also think this is tricky without type inference. @DetachHead I'm curious to know how did you run into this, do you have an example of a real world code base?

**Comment by DetachHead** [comment]
yeah there are about a hundred false positives in our codebase after enabling this rule. i haven't looked through all of them but it seems that most of them fall into one of the following scenarios:

- defining the type of an empty dict and populating it later:
  ```py
  foo = dict[str, int]()
  ...
  foo["bar"] = 1
  ```
  an obvious workaround is to just do `foo: dict[str, int] = {}` but nevertheless this is still some real world code
- defining the type of a `defaultdict` while instantiating it:
  ```py
  foo = defaultdict[str, list[str]](list)
  ```
- type aliases that don't use the `TypeAlias` annotation:
  ```py
  Foo = Literal["a", "b"]
  ```

all the examples i found use stdlib types so i guess in these cases they can be special-cased to reduce the risk of these false positives

**Comment by dhruvmanila** [comment]
Interesting, thanks for sharing that. cc @AlexWaygood do you think it's currently feasible to resolve this?

**Comment by AlexWaygood** [comment]
Hrm, this is hard!

For your three bullets in https://github.com/astral-sh/ruff/issues/13065#issuecomment-2306282918: bullets (1) and (2) would be easy with type inference; we could just check to see if the subscripted object was a type that had `__class_getitem__` defined.

Bullet point 3 (detecting whether an assignment not annotated using `TypeAlias` is intended to be a runtime value or a type alias) would be hard even with type inference. It was for this very reason that `typing.TypeAlias` (backported as `typing_extensions.TypeAlias`) was introduced.

I can see several options, none perfect:
1. Ignore the rule for call expressions when `lint.ruff.parenthesize-tuple-in-subscript = true`. This would reduce false positives but it wouldn't get rid of them, and would also introduce false negatives.
2. Special-case symbols from `builtins` and the `typing` module. And maybe other standard-library modules? It would reduce false positives, but again wouldn't get rid of them; there are lots of third-party generic types, after all.
3. Add a prominent warning to the docs that setting `lint.ruff.parenthesize-tuple-in-subscript = true` may lead to a lot of false positives if your code is heavily typed.

**Comment by BarrensZeppelin** [comment]
Hi, I can chime in with another data point.
In my code base I have 21 false positives from this rule.

* 14 instances are from creating empty dictionaries (or `defaultdict` / `WeakKeyDictionary`): `dict[int, str]()`
* 3 instances stem from converting a list to a tuple with known length: `tuple[int, int, int](my_list)`.
* 2 instances occur in interfacing with pydantic:
  `Adapter = TypeAdapter(dict[str, list[str]])`
  `MyID = Annotated[str, AfterValidator(_check_id)]`
* 1 instance occurs in the value for `default_factory` for a `dataclasses.field`:
  `_cache: Final = field(default_factory=dict[str, CachedObject], init=False)`
* 1 instance occurs when creating a type alias for a collection of literal strings without the `TypeAlias` annotation:
  `MyStrings = Literal["foo", "bar", "baz"]`

**Comment by dylwil3** [comment]
Maybe in the short term I can try to disable the rule for calls on builtin types like duct and tuple since I think there is some support for that. I also think it's worth the potential false negative to disable the rule for "Literal" since I doubt many people are shadowing that name and subscripting with it - plus `Literal` appears to be an instance where parentheses are explicitly disallowed, so it's a bit more dangerous of a false positive.

But I'll wait for the green light from the maintainers.

**Comment by dylwil3** [comment]
See https://peps.python.org/pep-0586/#illegal-parameters-for-literal-at-type-check-time

**Comment by MrRobotMan** [comment]
I'd like to chime in with one more datapoint. This issue comes up with dataclasses and default_factories.

```
@dataclass  
class MyClass:  
    items: dict[int, OtherClass] = field(default_factory=dict[int, "OtherClass"])  # RUF031 error
```
